src/ports/postgres/modules/utilities/utilities.py_in

```diff
@@ -775,6 +775,17 @@ def is_superuser(user):
     return plpy.execute("SELECT rolsuper FROM pg_catalog.pg_roles "\
                         "WHERE rolname = '{0}'".format(user))[0]['rolsuper']
 
+def get_table_owner(schema_table):
+
+    split_table = schema_table.split(".",1)
+    schema = split_table[0]
```

Comment:
I think we need to be more careful here.

Let's say there is a custom function table an admin created called `madlib.custom_functions`

I think I see a loophole in the way `get_table_ower` is implemented which allows any ordinary user to gain admin access.

Steps:
1.  User creates a table in public schema named "madlib.custom_functions.haha", filling it with their own malicious custom functions.
2.  User sets search_path=madlib,public
3.  Malicious user passes object_table='madlib.custom_functions.haha' to MADlib function

Seems like this would pass the security check, and then proceed to load the custom functions from the user's table instead of the admin's table.

Reply:
Oh, I guess these wouldn't actually work, because of the ,1 argument passed to split.
But still, it seems risky... maybe we should think it through carefully and add some more test cases to make sure.

What happens if the user table in the public schema is named "madlib.custom_functions"?

Some users may have permission to create their own schemas, but not have full admin access.  They could create a schema named "madlib.custom" and a table named "_functions"... seems like that might slip through, but I'd have to look at it more to be sure.

Reply:
There are two aspects at play. The accuracy of the `get_table_owner` utility function and how it is used in `load_custom_function`. 

Let's start with the second point. `load_custom_function` appends the madlib schema to the given parameter no matter what it looks like (madlib_keras_custom_function.py_in, line 71).  When we call the `get_table_owner`, our own prefixed schema name is split and the users' `object_table` argument is tested. This means the user cannot trick us with any schema name because we don't accept any and we don't parse their input, we directly look at the catalog to see if we have an exact match for the table name.

Getting back to the first point, the accuracy of the `get_table_owner` function. If some other future function tries to use this (which is possible since this is a generic utility function, not DL specific) there might be a way to break it since we might not have put the same amount of control on the input. I guess they might have a schema name like "as.df" which will mess up the split. Please note that the existing `get_schema` utility function uses a similar split. 

Considering these two points, I cannot think of an issue for `load_custom_function` unless the dbadmin installed madlib to a schema with `.` in its name. madpack does not allow usage of `.` in the schema name but an admin can rename the `madlib` schema once it is installed. I don't know if madlib even works in this state but I don't think we can account for any arbitrary manipulations of our installation.

Reply:
Ok, that makes sense!  I thought I remembered us only prepending madlib if they didn't already specify it... must have misremembered.  I'm not worried about the case where the admin installs madlib to a schema with a . in it, that seems like enough of a bizarre situation we can ignore it.

If we do want this to be a general utility function, then we might want to let postgres do the splitting into schema + table name rather than hoping we're doing it in exactly the same way it's done internally.  But I don't know off the top of my head how to do that (pg_catalog.pg_class?)